mmf/utils/logger.py

```diff
@@ -1,174 +1,251 @@
-        self.should_log = not should_not_log
+    if not PathManager.exists(log_folder):
+        PathManager.mkdirs(log_folder)
 
-        # Single log wrapper map
-        self._single_log_map = set()
+    if folder_only:
+        return log_folder
 
-    def add_handlers(self, channel: Type[logging.Handler]):
-        self._logger.addHandler(channel)
-        self._file_only_logger.addHandler(channel)
-        self._warnings_logger.addHandler(channel)
-        if self._stdout_logger:
-            self._stdout_logger.addHandler(channel)
+    log_filename = os.path.join(log_folder, log_filename)
 
-    def write(self, x, level="info", donot_print=False, log_all=False):
-        if self._logger is None:
-            return
+    return log_filename
 
-        if log_all is False and not self._is_master:
-            return
 
-        # if it should not log then just print it
-        if self.should_log:
-            if hasattr(self._logger, level):
-                if donot_print:
-                    getattr(self._file_only_logger, level)(str(x))
-                else:
-                    getattr(self._logger, level)(str(x))
-            else:
-                self._logger.error("Unknown log level type: %s" % level)
+# so that calling setup_logger multiple times won't add many handlers
+@functools.lru_cache()
+def setup_logger(
+    output: str = None,
+    color: bool = True,
+    name: str = "mmf",
+    disable: bool = False,
+    clear_handlers=True,
+    *args,
+    **kwargs,
+):
+    """
+    Initialize the MMF logger and set its verbosity level to "INFO".
+    Outside libraries shouldn't call this in case they have set there
+    own logging handlers and setup. If they do, and don't want to
+    clear handlers, pass clear_handlers options.
+
+    Args:
+        output (str): a file name or a directory to save log.
+            If ends with ".txt" or ".log", assumed to be a file name.
+            Default: Saved to file <save_dir/logs/log_[timestamp].txt>
+        color (bool): If false, won't log colored logs. Default: true
+        name (str): the root module name of this logger. Defaults to "mmf".
+        clear_handlers (bool): If false, won't clear existing handlers.
+
+    Returns:
+        logging.Logger: a logger
+    """
+    if disable:
+        return None
+    logger = logging.getLogger(name)
+    logger.propagate = False
+
+    logging.captureWarnings(True)
+    warnings_logger = logging.getLogger("py.warnings")
+
+    plain_formatter = logging.Formatter(
+        "%(asctime)s | %(levelname)s | %(name)s : %(message)s",
+        datefmt="%Y-%m-%dT%H:%M:%S",
+    )
+
+    distributed_rank = get_rank()
+    handlers = []
+
+    if distributed_rank == 0:
+        logger.setLevel(logging.INFO)
+        ch = logging.StreamHandler(stream=sys.stdout)
+        ch.setLevel(logging.INFO)
+        if color:
+            formatter = ColorfulFormatter(
+                colored("%(asctime)s | %(name)s: ", "green") + "%(message)s",
```

Comment:
Missing `%(levelname)s`

Reply:
Colorful Formatter has color for log levels. INFO is green, and when warning or error are logged they are appended on front in red color. Check the source for Colorful Formatter.